### PR TITLE
Email admin when a new application is received

### PR DESCRIPTION
**Pull request by sic-f**
PT Story: 
https://www.pivotaltracker.com/story/show/149374145

Changes proposed in this pull request:
1. Send email to admin when a new application is submitted

Screenshots (Optional):

Ready for review:
@patmbolger @weedySeaDragon @thesuss

### CODE REVIEW

File: app/controllers/application_controller.rb
Comment (weedySeaDragon) [inline app/controllers/application_controller.rb]: no need for this.  use the `admin?` method for the User model

File: app/controllers/application_controller.rb
Comment (RobertCram) [inline app/controllers/application_controller.rb]: The #admins method seems a better fit for the model than for the controller.

File: app/mailers/application_mailer.rb
Comment (RobertCram) [inline app/mailers/application_mailer.rb]: We have a ShfMailer base class that should probably used (See PR https://github.com/AgileVentures/shf-project/pull/318). This is the reason this PT story had a "blocker" (#140637017 must be approved and working). Blockers are indicated in PT by a red arrow in the lower left corner.

File: app/models/membership_application.rb
Comment (RobertCram) [inline app/models/membership_application.rb]: I don't think these delegates are necessary. The `:email` delegate doesn't seem to be used and at the moment the `:full_name` has already been provided for in the `MembershipApplicationsHelper` as `member_full_name`, and can be used in the mail template after adding `add_template_helper(MembershipApplicationsHelper)` to the mailer.

I think it's a matter of taste which approach for the getting full_name is better. I guess it depends on if you see it as a view-only property or not. I actually prefer your approach, but we shouldn't be using both.

File: app/models/user.rb
Comment (RobertCram) [inline app/models/user.rb]: See remark above. This has already been provided for in the `MembershipApplicationsHelper` as `member_full_name`

File: app/views/membership_application_mailer/new_application.html.haml
Comment (RobertCram) [inline app/views/membership_application_mailer/new_application.html.haml]: This should probably make use of translations and preferably make use of the ShfMailer templates to encourage a uniform look and feel for all emails.  

The `#{@membership_application.user_full_name}` can be replaced with `#{member_full_name}` after adding the `MembershipApplicationsHelper` to the mailer class (see above)

**Review by RobertCram — CHANGES_REQUESTED** [review]

File: app/controllers/membership_applications_controller.rb
Comment (weedySeaDragon) [inline app/controllers/membership_applications_controller.rb]: This logic shouldn't go under the "if there is a new file uploaded..." condition.  It should be it's own small, simple method, and should be called after this conditional structure.

As Robert noted, once #318  is accepted, then you can use the `ShfMailer.`  Let `ShfMailer` have the responsibility for when and how to deliver: define a method in ShfMailer to send the mail, and put all of the logic for sending it in that method.  (Don't use `deliver_later` in this controller; keep the responsibility within the mailer used))

File: app/models/membership_application.rb
Comment (weedySeaDragon) [reply app/models/membership_application.rb]: In #318 I added a method to `User` to take care of this.

File: app/models/user.rb
Comment (weedySeaDragon) [reply app/models/user.rb]: I agree with this approach, @RobertCram  -- I think that the method does belong in `User`  I think that we should change the code in `MembershipApplicationsHelper  .member_full_name` to now just call this method.

**Review by weedySeaDragon — CHANGES_REQUESTED** [review]

File: app/models/user.rb
Comment (RobertCram) [reply app/models/user.rb]: @Ashley, If we go for making `full_name` a user attribute (which I agree with), I don't see any advantage to having a `member_full_name` in the `MembershipsApplicationHelper`. A refactoring to a `full_name` delegate on `MembershipApplication` would then seem a better fit.

But that would mean changes to some tests, views and helpers that have actually nothing to do with this feature. So I think that refactoring would probably be best done in a separate chore.

File: app/models/membership_application.rb
Comment (weedySeaDragon) [reply app/models/membership_application.rb]: After talking with Robert after our dev team meeting on Thursday, I agree with you both that we should use `delegate :full_name, to :user ...`

File: app/views/membership_application_mailer/new_application.html.haml
Comment (sic-f) [reply app/views/membership_application_mailer/new_application.html.haml]: @RobertCram will have a look at the translations and ShfMailer templates. Am I right that we are using `@membership_application.user_full_name` now?

File: app/controllers/application_controller.rb
Comment (sic-f) [reply app/controllers/application_controller.rb]: @RobertCram ok will transfer this to `User`

File: app/models/membership_application.rb
Comment (sic-f) [reply app/models/membership_application.rb]: @RobertCram @weedySeaDragon so we'll go with delegating `#full_name` to `User`? TBH I can't even remember why I added the `#email` delegate to `User`. I will just delete it.

File: app/models/user.rb
Comment (sic-f) [reply app/models/user.rb]: @RobertCram @weedySeaDragon ok I will delete `#full_name` in lines `10-12` in `User`

File: app/controllers/membership_applications_controller.rb
Comment (sic-f) [reply app/controllers/membership_applications_controller.rb]: @weedySeaDragon ok will just wait until #318 is merged.

File: app/mailers/application_mailer.rb
Comment (sic-f) [reply app/mailers/application_mailer.rb]: @RobertCram noted. I will use it once #318 is merged.

File: app/models/membership_application.rb
Comment (sic-f) [reply app/models/membership_application.rb]: @RobertCram @weedySeaDragon keeping `#full_name` delegate and removing `#email` delegate

File: app/views/membership_application_mailer/new_application.html.haml
Comment (RobertCram) [reply app/views/membership_application_mailer/new_application.html.haml]: @sic-f I think your best bet is to:

1) Wait for PR https://github.com/AgileVentures/shf-project/pull/318 to be merged (will be soon)
2) Implement PT https://www.pivotaltracker.com/story/show/149932678 (and have it merged)
3) Rebase this PR to develop, and then implement the requested changes.

Does that make sense?

File: app/controllers/application_controller.rb
Comment (sic-f) [reply app/controllers/application_controller.rb]: @weedySeaDragon This method was for fetching all the admins. I've moved this to `User` so I can do `User.admins` in the controller. Please correct me if I have misunderstood. Thanks.

File: app/views/membership_application_mailer/new_application.html.haml
Comment (sic-f) [reply app/views/membership_application_mailer/new_application.html.haml]: @RobertCram yes that makes sense. Will wait for #318 to get merged first.

File: app/controllers/membership_applications_controller.rb
Comment (thesuss) [reply app/controllers/membership_applications_controller.rb]: @sic-f  Now this seems to be the case. =)

Comment (weedySeaDragon) [reply app/controllers/membership_applications_controller.rb]: @sic-f -- Now this is ready to be worked on. The PR that was actually merged was #359 .

 You should probably close this one and open a new PR since so much of the code base has changed.  You will also need to:
- get the exact text and formatting needed for the emails 

Doing this should now be really easy, based on the work for #359.

Comment (sic-f) [reply app/controllers/membership_applications_controller.rb]: @thesuss @weedySeaDragon I have reworked this PR. Ready to be reviewed again.

File: app/controllers/membership_applications_controller.rb
Comment (weedySeaDragon) [inline app/controllers/membership_applications_controller.rb]: This shouldn't go in this conditional.  We should send the email even if `new_file_uploaded(params)` fails.
Put this after the conditional

File: app/controllers/membership_applications_controller.rb
Comment (weedySeaDragon) [inline app/controllers/membership_applications_controller.rb]: We're not using `.deliver_later` since the mailer is done via HTTP (REST) calls to the mailgun mailer.

File: app/controllers/membership_applications_controller.rb
Comment (weedySeaDragon) [inline app/controllers/membership_applications_controller.rb:270]: This is good.  We _should_ be sending email out to every user that is an admin.  This means that we need to make sure that we use the address for each admin as the @recipient address in the email that is sent out and _not_ use the generic ENV['SHF_MEMBERSHIP_EMAIL'] address.  (The ENV['SHF_MEMBERSHIP_EMAIL'] is used as the return/sender address when sending some emails out.)


1.  Need to have a specification for this (= RSpec test for the mailer )
2. will need to write code in the AdminMailer to handle this

(You need to have ENV['SHF_MEMBERSHIP_EMAIL']  in your .env but while you're just testing, it shouldn't matter what the value is.  The `.env.example` file shows you all of the values that you need (with dummy data).  Use it to compare to your `.env` file to see what keys you might be missing.)

**Review by weedySeaDragon — CHANGES_REQUESTED** [review]
You need to fix the previewer.  Your previewer is trying to call the AdminMailer with the wrong number of arguments:

```
AdminMailer.member_application_received(random_member_app)
```
needs a 2nd argument for the admin.  I suggest that before you call this in the previewer, just grab an admin user from the db.  Something like `   some_admin = User.find_by(admin: true)` and pass that in to the method.

Also -- need a feature for this.

File: app/controllers/membership_applications_controller.rb
Comment (sic-f) [reply app/controllers/membership_applications_controller.rb:270]: @weedySeaDragon so can I safely change `ENV['SHF_MEMBERSHIP_EMAIL']` in the private method

```
  def set_recipient_email(_record)
    @recipient_email = ENV['SHF_MEMBERSHIP_EMAIL']
  end
```

in `AdminMailer` to  `@recipient_email = record.email if record.respond_to? :email`?

File: app/controllers/membership_applications_controller.rb
Comment (sic-f) [reply app/controllers/membership_applications_controller.rb]: @weedySeaDragon I used `deliver_now`. If this is still wrong please let me know.

File: app/controllers/membership_applications_controller.rb
Comment (sic-f) [reply app/controllers/membership_applications_controller.rb]: @weedySeaDragon put this after the conditional.

File: app/controllers/membership_applications_controller.rb
Comment (weedySeaDragon) [inline app/controllers/membership_applications_controller.rb:59]: There are 2 **distinct** emails that need to sent out. This is documented in the wiki: https://github.com/AgileVentures/shf-project/wiki/Emails-sent-when-a-Membership-Application-changes

Don't delete the code that sends out the acknowledgment.  You need to _**add**_ a line that will send out emails to admins.

File: app/controllers/membership_applications_controller.rb
Comment (weedySeaDragon) [reply app/controllers/membership_applications_controller.rb:270]: No.  ENV['SHF_MEMBERSHIP_EMAIL'] is used for a different purpose. It is used as the single email address that members can use (email to) if they have questions about their membership. 

The recipients for your emails are the User.admins.

File: app/mailers/admin_mailer.rb
Comment (weedySeaDragon) [inline app/mailers/admin_mailer.rb]: **`AdminMailer` should be a subclass of `ApplicationMailer`.**  The purpose of the class `AbstractMembershipInfoMailer` is to send emails to members.  The purpose of `AdminMailer` should be to send emails to admins (not members).

File: app/mailers/abstract_membership_info_mailer.rb
Comment (weedySeaDragon) [inline app/mailers/abstract_membership_info_mailer.rb:26]: These methods don't belong here.  They should go into the `AdminMailer` class because they're only about sending emails to admins, not members.  (See my comment below about the AdminMailer class)

These can be combined into 1 method.  I don't see how separating them can help in this case (once they're in the AdminMailer class).

File: spec/mailers/admin_mailer_spec.rb
Comment (weedySeaDragon) [inline spec/mailers/admin_mailer_spec.rb]: You need to specify what the greeting should be:
` I18n.t('application_mailer.greeting', greeting_name: 'Firstname Lastname')

File: spec/mailers/admin_mailer_spec.rb
Comment (weedySeaDragon) [inline spec/mailers/admin_mailer_spec.rb]: You need to add specification for ` it 'has a previewer' ...` (use the spec in the `membership_application_spec.rb` as an example)

You need to add specifications for the other important parts of the email so that future maintainers understand that the information is required in the email, and so that it can serve as tests to ensure that this and future code is correct.  Ex:
- it has the correct  member_app.user.full_name
- it has the correct member_app.contact_email
- it has the correct url_for(@member_app)
- it has the text for the new app has arrived (the t('....'))
- it has the text for the view app here (the t('....'))
- it has the text for must be logged in (the t('....'))

**Review by weedySeaDragon — CHANGES_REQUESTED** [review]
The RSpec specification "email us with questions shows membership email address from ENV['SHF_MEMBERSHIP_EMAIL']"
is failing because the email doesn't have that line in it.  (it should)

The scenario "User submits a new application and email is sent" in the feature "" is now failing.  It's failing because you removed the functionality.  See my comment in  membership_application_controller.rb.
This feature should pass; your work shouldn't affect it.
All RSpec and Cucumber features should be passing.  If your work is still in progress and you know that there are tests that are failing, mark your PR as a WIP.

I've created a Cucumber Feature that describes what should happen overall.  Add
the following should be put into a file called `features/emails/new-application-email-to-admins.feature` : 

```
Feature: When a new application is received, all admins get an email notification

  As an admin,
  So that I know that SHF received a new application and that we need to start reviewing it,
  I should get an email with a link to the application so I can go look at it
  and the email of the applicant so I have it if there is some problem with the application or the SHF system
  I can contact the applicant easily.


  Background:

    Given the following users exists
      | email               | admin |
      | admin1@shf.se       | yes   |
      | admin2@shf.se       | yes   |
      | emma@happymutts.com |       |


    And the following business categories exist
      | name         |
      | Groomer      |


  Scenario: User submits a new application and email is sent to all admins
    Given I am logged in as "emma@happymutts.com"
    And I am on the "landing" page
    And I click on t("menus.nav.users.apply_for_membership")
    And I fill in the translated form with data:
      | membership_applications.new.first_name | membership_applications.new.last_name | membership_applications.new.company_number | membership_applications.new.phone_number | membership_applications.new.contact_email |
      | Emma                                   | HappyMutts                            | 5562252998                                 | 031-1234567                              | emma@happymutts.com                       |
    And I select "Groomer" Category
    And I click on t("membership_applications.new.submit_button_label")
    Then I should be on the "landing" page
    And I should see t("membership_applications.create.success", email_address: 'emma@happymutts.com')
    And I am logged out
    And I am logged in as "admin1@shf.se"
    Then "admin1@shf.se" should receive an email
    And I open the email
    And I should see t("application_mailer.admin.new_application_received.subject") in the email subject
    And I should see t("application_mailer.admin.new_application_received.message_text.new_app_arrived") in the email body
    And I should see t("application_mailer.admin.new_application_received.message_text.from") in the email body
    And I should see t("application_mailer.admin.new_application_received.message_text.view_app_here") in the email body
    And I should see t("application_mailer.admin.new_application_received.message_text.must_be_logged_in") in the email body
    And I am logged out
    And I am logged in as "admin2@shf.se"
    Then "admin2@shf.se" should receive an email
    And I open the email
    And I should see t("application_mailer.admin.new_application_received.subject") in the email subject
    And I should see t("application_mailer.admin.new_application_received.message_text.new_app_arrived") in the email body
    And I should see t("application_mailer.admin.new_application_received.message_text.from") in the email body
    And I should see t("application_mailer.admin.new_application_received.message_text.view_app_here") in the email body
    And I should see t("application_mailer.admin.new_application_received.message_text.must_be_logged_in") in the email body


  Scenario: User submits a new application app with bad info so it is not created, so no email sent [SAD PATH]
    Given I am logged in as "emma@happymutts.com"
    And I am on the "landing" page
    And I click on t("menus.nav.users.apply_for_membership")
    And I fill in the translated form with data:
      | membership_applications.new.first_name | membership_applications.new.phone_number | membership_applications.new.contact_email |
      | Emma                                   | 031-1234567                              | emma@happymutts.com                       |
    And I select "Groomer" Category
    And I click on t("membership_applications.new.submit_button_label")
    And I should see t("membership_applications.create.error")
    Then "admin1@shf.se" should receive 0 email
    Then "admin2@shf.se" should receive 0 email

```

File: app/controllers/membership_applications_controller.rb
Comment (sic-f) [reply app/controllers/membership_applications_controller.rb:270]: @weedySeaDragon Thanks. Although this one I think is for the admins' emails? 

So is it right in my understanding that `ENV['SHF_MEMBERSHIP_EMAIL`] is "SHF's email" address and would be wrong to send to it for this purpose? 

And that `@recipient_email = record.email if record.respond_to? :email` would be the correct one since this code fetches an admin's email?

And actually I have deleted the private methods in `AdminMailer`, because they are duplicates of those in `CommonMailUtils`?

Although the `#set_recipient_email` in `AdminMailer` is a bit different from the one in `CommonMailUtils` because it sets `ENV['SHF_MEMBERSHIP_EMAIL]` as the recipient, which does not make sense to me since that is not an admin's email?

File: app/controllers/membership_applications_controller.rb
Comment (sic-f) [reply app/controllers/membership_applications_controller.rb:59]: @weedySeaDragon oh I'm sorry. I thought `MembershipApplicationMailer.acknowledge_received(@membership_application).deliver` was also my code that I wrote in the past that did not apply anymore. I did not notice it was someone else's. I will put it back. Thank you for spotting it.

Comment (sic-f) [reply app/controllers/membership_applications_controller.rb:59]: @weedySeaDragon "The RSpec specification "email us with questions shows membership email address from ENV['SHF_MEMBERSHIP_EMAIL']"
is failing because the email doesn't have that line in it. (it should)"

Looking at the email template for the `AdminMailer`, the line for specifying the `ENV['SHF_MEMBERSHIP_EMAIL']` is not appropriate since we are emailing admins and not members. So there is no need to add that line to the email template?

Would that also mean we can't use the shared emailed test for the `AdminMailer` spec since the shared examples include adding that line with the `ENV['SHF_MEMBERSHIP_EMAIL`]?

File: app/controllers/membership_applications_controller.rb
Comment (weedySeaDragon) [reply app/controllers/membership_applications_controller.rb:270]: ENV['SHF_MEMBERSHIP_EMAIL] is the email address that people (members and potential members) can send to when they have questions about SHF members.  It is not the 'general SHF email address.' 

It is _not_ a recipient for this email.

The code you have in `send_new_membership_application_notice_to_admins` is the right way to identify and gather the recipients for the email.

> And that @recipient_email = record.email if record.respond_to? :email would be the correct one since this code fetches an admin's email?

I don't know that you mean by this.

>And actually I have deleted the private methods in AdminMailer, because they are duplicates of those in CommonMailUtils?

I don't see this change in your code.  You'll need to push your commits here so that I can see what you mean.

File: app/controllers/membership_applications_controller.rb
Comment (weedySeaDragon) [reply app/controllers/membership_applications_controller.rb:59]: Let's leave that line in the email for right now.  We'll review that with the client.  (My thought is that people that are new SHF admins may have questions and having that info in the email for them could be handy.  But the client can give us feedback on that idea and let us know.)

If we need to take it out, then yes, we'll need to change the test for the emails:  we may want to refactor the `shared_example` test.  But let's wait and get feedback from the client first.

**Comment by thesuss** [comment]
I'd like to see some screen-shots in the description if possible 💚

**Review by weedySeaDragon — APPROVED** [review]

File: app/controllers/membership_applications_controller.rb
Comment (weedySeaDragon) [reply app/controllers/membership_applications_controller.rb:59]: still needed

**Review by weedySeaDragon — CHANGES_REQUESTED** [review]

**Comment by weedySeaDragon** [comment]
this was done with PR #415

**Pull request closed without merging**
